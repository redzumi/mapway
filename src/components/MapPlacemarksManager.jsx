import React, { Component } from 'react'
import shortid from 'shortid'
import styled from 'styled-components'

import AddPlacemarkForm from './AddPlacemarkForm'
import SortablePlacemarksList from './SortablePlacemarksList'
import YandexMapContainer from './YandexMapContainer'

const DEFAULT_MAP_CENTER = [55.75120138894101, 37.62371955158138] // Moscow

class MapPlacemarksManager extends Component {
  constructor(props) {
    super(props)
    this.state = {
      placemarks: [],
      mapCenter: DEFAULT_MAP_CENTER,
    }
  }

  handlePlacemarkCoordinatesUpdate = (index, coordinates) => {
    const placemarks = this.state.placemarks.slice()
    placemarks[index] = { ...placemarks[index], coordinates }
    this.setState({ placemarks })
  }

  handlePlacemarkAdd = placemark => {
    this.setState({
      placemarks: [
        ...this.state.placemarks,
        { id: shortid.generate(), coordinates: this.state.mapCenter, ...placemark },
      ],
    })
  }

  handlePlacemarksOrderUpdate = placemarks => {
    this.setState({ placemarks })
  }

  handlePlacemarkDelete = index => {
    const placemarks = this.state.placemarks.slice()
    placemarks.splice(index, 1)
    this.setState({ placemarks })
  }

  handleMapCenterUpdate = mapCenter => {
    this.setState({ mapCenter })
  }

  render() {
    return (
      <StyledContainer>
        <LeftPanel>
          <AddPlacemarkForm onPlacemarkAdd={this.handlePlacemarkAdd} />
          <SortablePlacemarksList
            placemarks={this.state.placemarks}
            onPlacemarksOrderUpdate={this.handlePlacemarksOrderUpdate}
            onPlacemarkDelete={this.handlePlacemarkDelete}
          />
        </LeftPanel>
        <YandexMapContainer
          placemarks={this.state.placemarks}
          onPlacemarkCoordinatesUpdate={this.handlePlacemarkCoordinatesUpdate}
          mapCenter={this.state.mapCenter}
          onMapCenterUpdate={this.handleMapCenterUpdate}
        />
      </StyledContainer>
    )
  }
}

const LeftPanel = styled.div`
  padding: 10px;
`

const StyledContainer = styled.div`
  display: flex;
`

export default MapPlacemarksManager
