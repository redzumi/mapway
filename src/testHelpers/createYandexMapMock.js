const createYandexMapMock = onMapInitializationFinished => {
  class EventsCollection {
    add = () => {}
  }

  class GeoObjectCollection {
    constructor() {
      this.events = new EventsCollection()
    }

    add = () => {}
    set = () => {}
    getLength = () => {}
    each = () => {}
    splice = () => {}
  }

  class OptionsCollection {
    get = () => this
    fromGlobalPixels = () => 'test'
  }

  class LineString {
    set = () => {}
    splice = () => {}
  }

  class Geometry {
    LineString = () => new LineString()
    getCoordinates = () => {}
  }

  class Polyline {}

  class GeoObject {
    constructor() {
      this.geometry = new Geometry()
      this.events = new EventsCollection()
      this.properties = {
        get: () => {},
        set: () => {},
      }
      this.hint = {
        open: () => {},
      }
    }
  }

  class Map {
    constructor() {
      this.geoObjects = new GeoObjectCollection()
      this.events = new EventsCollection()
      this.options = new OptionsCollection()
    }
  }

  const geocode = async () => ({
    geoObjects: {
      get: () => ({
        getAddressLine: () => 'Test Address',
      }),
    },
  })

  return {
    ready: callback => {
      callback()
      onMapInitializationFinished()
    },
    geometry: new Geometry(),
    geocode,
    Map,
    GeoObject,
    GeoObjectCollection,
    Polyline,
  }
}

export default createYandexMapMock
