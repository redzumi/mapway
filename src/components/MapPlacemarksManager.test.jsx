import React from 'react'
import { mount } from 'enzyme'
import createYandexMapMock from '../testHelpers/createYandexMapMock'
import MapPlacemarksManager from './MapPlacemarksManager'

let managerComponent
let managerInstance
beforeEach(done => {
  window.ymaps = createYandexMapMock(done)
  managerComponent = mount(<MapPlacemarksManager />)
  managerInstance = managerComponent.instance()
})

afterEach(() => {
  managerComponent.unmount()
})

describe('MapPlacemarksManager', () => {
  it('renders correctly', () => {
    expect(managerComponent).toMatchSnapshot()
  })

  it('handles placemark coords updating', () => {
    managerInstance.setState({
      placemarks: [{ id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] }],
    })
    managerInstance.handlePlacemarkCoordinatesUpdate(0, [55, 55])

    expect(managerComponent.state().placemarks[0]).toEqual({
      id: 'id1',
      name: 'Test placemark 1',
      coordinates: [55, 55],
    })
  })

  it('handles placemark addition', () => {
    managerInstance.setState({ mapCenter: [22, 22] })
    managerInstance.handlePlacemarkAdd({ name: 'Test placemark 1' })

    expect(managerComponent.state().placemarks[0]).toEqual({
      name: 'Test placemark 1',
      id: expect.anything(),
      coordinates: [22, 22],
    })
  })

  it('handles placemark reordering', () => {
    managerInstance.setState({
      placemarks: [
        { id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] },
        { id: 'id2', name: 'Test placemark 2', coordinates: [23, 23] },
      ],
    })

    managerInstance.handlePlacemarksOrderUpdate([
      { id: 'id2', name: 'Test placemark 2', coordinates: [23, 23] },
      { id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] },
    ])

    expect(managerComponent.state().placemarks).toEqual([
      { id: 'id2', name: 'Test placemark 2', coordinates: [23, 23] },
      { id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] },
    ])
  })

  it('handles placemark deletion', () => {
    managerInstance.setState({
      placemarks: [
        { id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] },
        { id: 'id2', name: 'Test placemark 2', coordinates: [23, 23] },
      ],
    })

    managerInstance.handlePlacemarkDelete(1)

    expect(managerComponent.state().placemarks).toEqual([
      { id: 'id1', name: 'Test placemark 1', coordinates: [22, 22] },
    ])
  })

  it('handles map center updating', () => {
    managerInstance.setState({ mapCenter: [22, 22] })
    managerInstance.handleMapCenterUpdate([55, 55])

    expect(managerComponent.state().mapCenter).toEqual([55, 55])
  })
})
