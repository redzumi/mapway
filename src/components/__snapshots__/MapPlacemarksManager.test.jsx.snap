// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapPlacemarksManager renders correctly 1`] = `
.c2 {
  background-color: #fff;
  width: 300px;
  font-family: 'Montserrat',sans-serif;
  padding-left: 10px;
  height: 60px;
  margin: 0px;
  outline: none;
  box-sizing: border-box;
  border: 1px solid #efefef;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  margin-bottom: 20px;
}

.c2:focus {
  border: 1px solid #a7a7a7;
}

.c3 {
  background-color: #f3f3f3;
  border: 1px solid #efefef;
  width: 300px;
  padding: 0px;
  margin: 0px;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 li {
  box-shadow: none;
}

.c3:active {
  cursor: row-resize;
}

.c4 {
  width: 600px;
  height: 400px;
  padding: 10px;
}

.c1 {
  padding: 10px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<MapPlacemarksManager>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          <AddPlacemarkForm
            onPlacemarkAdd={[Function]}
          >
            <form
              onSubmit={[Function]}
            >
              <styled.input
                name="name"
                onChange={[Function]}
                placeholder="Новая точка маршрута"
                required={true}
                value=""
              >
                <input
                  className="c2"
                  name="name"
                  onChange={[Function]}
                  placeholder="Новая точка маршрута"
                  required={true}
                  value=""
                />
              </styled.input>
            </form>
          </AddPlacemarkForm>
          <SortablePlacemarksList
            onPlacemarkDelete={[Function]}
            onPlacemarksOrderUpdate={[Function]}
            placemarks={Array []}
          >
            <sortableList
              axis="y"
              distance={0}
              getHelperDimensions={[Function]}
              hideSortableGhost={true}
              items={Array []}
              lockAxis="y"
              lockOffset="50%"
              lockToContainerEdges={true}
              onItemDelete={[Function]}
              onSortEnd={[Function]}
              pressDelay={0}
              pressThreshold={5}
              shouldCancelStart={[Function]}
              transitionDuration={300}
              useDragHandle={true}
              useWindowAsScrollContainer={false}
            >
              <Component
                items={Array []}
                onItemDelete={[Function]}
              >
                <styled.ul>
                  <ul
                    className="c3"
                  />
                </styled.ul>
              </Component>
            </sortableList>
          </SortablePlacemarksList>
        </div>
      </styled.div>
      <YandexMapContainer
        mapCenter={
          Array [
            55.75120138894101,
            37.62371955158138,
          ]
        }
        onMapCenterUpdate={[Function]}
        onPlacemarkCoordinatesUpdate={[Function]}
        placemarks={Array []}
      >
        <styled.div
          innerRef={[Function]}
        >
          <div
            className="c4"
          />
        </styled.div>
      </YandexMapContainer>
    </div>
  </styled.div>
</MapPlacemarksManager>
`;
